Tutorial Series

ScottWe opened the pull request:
Testing pull request between own branches.

Commit:
Fixed typos and commented code (testing pr).

Commit:
Tutorial(3): Added monitor and formal spec.

Commit:
Tutorial(3): streamlined mnaual instrumentation, and referred to monitor.

Commit:
Fixed function naming conventions.

Commit:
Tutorial(3): applied feedback on ghost variables.

Commit:
Tutorial(3): added regex with pred. language.

- The regex specifies the sequence of transactions taken by the monitor.
- The predicate language gives precisely how we abstract transactions.

Also switched from tex to graphviz for DFA.

Commit:
Tutorial(3): renamed post.

Commit:
Tutorial(3): Updated intro wrt goals.

Solidity only supports require and asserts.
This doesn't scale to complex properties and a spec lang would be beneficial.
We show how SmartACE could accomplish this, and work through an example mechanically.

Commit:
Tutorial(3): Removed prop and cex from contract section.

I now motivate the property, and suggest that a counterexample might exist.

Commit:
Tutorial(3): Revised property section. Switch to svg.

Commit:
Tutorial(3): Revised conclusion and logged owner.

Commit:
Tutorial(3): split model from instrumentation.

Commit:
Tutorial(2): separated out instrumented contract.

Commit:
Tutorial(4,wip): explains tut3's address count.

This tutorial is going to (1) introduce local reasoning in the case of
no mappings, (2) explain how SmartACE automates this, and (3) show how
to construct local proofs from seahorn proof cerificates.

This revision completes (1).

Commit:
Tutorial(3,fix): added missing instrumentation.

Commit:
Tutorial(4,wip): added invr interpretation

Commit:
Tutorial(4,wip): topology for local reasoning.

Described our assumptions on topology for local reasoning. Described the
relevant syntactic checks we can do.

Commit:
Tutorial(4,wip): explained restricted addresses.

Commit:
Tutorial(4): continouity fixes

Commit:
Tutorial(3,4): clarified commands, notation.

Commit:
Tutorial(5,wip): contract and prop w/o description

Commit:
Tutorial(5,wip): characterized network topology

Commit:
Tutorial(4,wip): revised: Limiting Addresses in...

Commit:
Tut(4): replaced equiv. class with example

Commit:
Tut(4): clarified syntax-directed client analysis.

Commit:
Tut(5,wip): described goal, contract, and property

Commit:
Tut(5,wip): introduced comp. invar.

Commit:
Tut(5): distinction between rep and interference.

I have added an open question on renaming the
representative and interference addresses. Those
names seem misleading.

Commit:
Tut(5,wip): rep addrs continued.

Commit:
Tut(5,wip): described comp. and adequacy checks.

Commit:
Tut(5,wip): described mapping encoding.

agurfinkel commented on tutorials/3_transactions.md:
```diff
@@ -1,27 +1,29 @@
+temporal logic, converted into monitors, and then instrumented mechanically.
+This instrumentation will be automated in future versions of SmartACE.
 
 This tutorial requires [Seahorn and clang-format](1_installation.md).
```
remove

Commit:
Tut(5,wip): instrumentation overview.

agurfinkel commented on tutorials/3_transactions.md:
**build targets** instead of **make targets**

agurfinkel commented on tutorials/3_transactions.md:
```diff
@@ -83,14 +130,16 @@ Let's take a look at the artifacts generated by the tool. You should see:
 
   * `mdkir build ; cd build`
   * `cmake .. -DSEA_PATH=/path/to/seahorn/bin`
   * `make run-clang-format`
```
replace with (but check) `cmake --build . --target clang-format`

this will work independently of how build utility is called

agurfinkel commented on tutorials/3_transactions.md:
```diff
@@ -83,14 +130,16 @@ Let's take a look at the artifacts generated by the tool. You should see:
 
-Now open `cmodel.c`. At the top we can see how each contract is encoded:
+We highlight four parts of the model: the contract encoding, the method
+encoding, the bundle construction, and the "transaction loop". We start at line
```
For those interested in the model, but do not want to run the tools, the full generated model is available `here`. The line numbers below refer to that model.

agurfinkel commented on tutorials/3_transactions.md:
```diff
@@ -99,127 +148,139 @@ struct Fund {
+
+`called_openFund` tracks when `Fund.openFund()` is called. `pre_Fund_balance`
+and `post_Fund_balance` store the values of `BALANCE(FUND)` before and after
+each transaction.  For now `GHOST_VAR` is a placeholder, but will be used for
```
remove **will be used**. Replace by,

note that `GHOST_VAR` is just a placeholder for readability and is ignored by the C compiler.

agurfinkel commented on tutorials/3_transactions.md:
````diff
@@ -99,127 +148,139 @@ struct Fund {
 
-(*SW: I suggest `CONTRACT_Method(_For_OVERRIDE)_METHODNAME`, `CONTRACT_INIT`, and `CONTRACT_FALLBACK`. Modifiers will come before `_METHODNAME`.*).
+```cpp
+GHOST_VAR int called_openFund;
````
I prefer to name global variables with `g_` prefix, so this is `g_called_openFund`.
Here, nicely, `g_` stands both for **ghost** and **global**

agurfinkel commented on tutorials/3_transactions.md:
```diff
@@ -99,127 +148,139 @@ struct Fund {
-## Debugging the Model
-
-When `Seahorn` detects that an assertion can be violated, it can generate a
+If we run `make verify` on the instrumented contract, we will see that the
```
replace `make verify` with `cmake --build . --target verify`

agurfinkel replied:
replace all uses of `make` with `cmake --build`. but please check the syntax

agurfinkel reviewed (commented):
Tutorials up to version 3 are ready to go. I've minor comments, but everything is great. Can we release them by Thu? What is needed are the docker container + place model in some github repo + validate the changes in commands that I am asking for.

agurfinkel commented on tutorials/4_arbitrary_clients.md:
add citation to R&N TACAS 2016 and mine FSE 2016

agurfinkel commented on tutorials/4_arbitrary_clients.md:
here and elsewhere, link to prior tutorials if they are mentioned

agurfinkel commented on tutorials/4_arbitrary_clients.md:
use `cmake --build`

put the model online. create github repo with all files used by the tutorial and point there.

agurfinkel commented on tutorials/4_arbitrary_clients.md:
`bitcode` not `bytcode`. here and elsewhere

agurfinkel commented on tutorials/4_arbitrary_clients.md:
use `==>` for implication

agurfinkel reviewed (commented):
tutorial 4 is very good and is ready to be published.

let's aim at publishing it by Thu along with tutorial 1-4!

Commit:
Tut(5,wip): instrumenting the property.

Commit:
Tut(5,wip): in-code example of comp invar.

wuestholz commented on tutorials/3_transactions.md:
```diff
@@ -99,127 +148,139 @@ struct Fund {
+wrapped by singleton c-struts in `primitives.h`. Members prefixed by `user_`
+are taken from the smart contract definition, while members prefixed by `model_`
+are generated by SmartACE. If a contract is constructed with `new`, it is nested
+inside the contract which calls `new`. To illustrate, if we defined a contract
```
I don't quite understand the nesting for dynamically constructed contracts. Why isn't the type of `user_fund` a pointer to the struct you allocate when constructing the new fund?

wuestholz replied:
Is this implicitely encoding some sort of ownership?

ScottWe replied:
There is an explicit assumption that's we're encoding a fixed number of smart contracts.

There's an implicit assumption (presently) that the number is fixed by restricting `new` to construction, and by requiring a 1-to-1 mapping between state variables and `new` calls.

This could be taken as ownership.

I've thought about relaxing some of the implicit assumptions but we haven't hit an example where it mattered.

I can elaborate on this in the post.

wuestholz replied:
I see. I wasn't aware of the assumption about encoding a fixed number of contracts. Does the tool enforce it by rejecting programs that allocate new contracts outside constructors?

I wonder if we can relax the assumption similar to how we use interference for user addresses.

I wouldn't elaborate on it in the post. In fact, I'd probably be more vague about this since it isn't crucial for understanding the post.

ScottWe replied:
That's essentially what we do.

I have had a few ideas on how to relax this, driven by examples we've come across seen.

1. Some smart contracts go through stages, and each stage is associated with a newly constructed helper contract. The number of stages in these contracts are usually bounded, so we could still do this case statically.
2. Some contracts take in addresses, and then cast those addresses to contracts (I believe this is what you're referencing). In this case, I had the same idea you were suggesting.

I'll reword it to imply "in this specific case, this is what we do."

ScottWe replied:
I looked at the wording again. It should have said "If a state variable of the contract type is initialized once by a call to `new`, ...". I think it's okay now.

wuestholz replied:
Sounds good. Yeah, (2) is definitely quite common (e.g., for tokens that are used by a contract). I was thinking that even for a dynamically created contract (with unbounded number of creations) one could use an "interference contract" to summarize the behavior.

wuestholz commented on tutorials/3_transactions.md:
```diff
@@ -99,127 +148,139 @@ struct Fund {
+notice that we restrict addresses to 3 and 4. The importance of this choice, and
+its impact on completeness, are the topic of the next tutorial. For now we can
+think of this as a bounded model with two clients, so at the least, our
+verification is sound.
```
When I read this it seems surprising that the verification would be sound if you have a bounded model.

ScottWe replied:
Good point. I might have soundness and completeness flipped here.

I was thinking wrt "the proof that a violation exists". In this context, soundness is "all detected violations are true violations" and completeness is "all true violations are detected."

I can flip this.

wuestholz replied:
👍 Yeah, these terms are sometimes used differently in different communities. However, in verification it's usually the other way around.

wuestholz commented on tutorials/3_transactions.md:
```diff
@@ -241,28 +302,59 @@ sender [uint8]: 3
+new fund, whose owner is `Manager`. This aligns with the first owner. Client 3
+then `claim()` and takes ownership. This aligns with the second owner. From
+here, client 3 is free to `open()` the contract, and then `deposit()` a single
+Ether. The balance of `Fund` has changed without a call to `openFund()`.
```
Maybe it would be good to somehow summarize each blog post at the end (2-3 sentences about what we "learned") and provide a pointer to what will be covered in subsequent posts.

wuestholz commented on tutorials/3_transactions.md:
```diff
@@ -241,28 +302,59 @@ sender [uint8]: 3
+We can now see how this is possible. When `Manager` is constructed, it creates a
+new fund, whose owner is `Manager`. This aligns with the first owner. Client 3
+then `claim()` and takes ownership. This aligns with the second owner. From
+here, client 3 is free to `open()` the contract, and then `deposit()` a single
```
It might be good to describe how one would fix this.

wuestholz commented on tutorials/4_arbitrary_clients.md:
Why 6 instead of 5?

ScottWe replied:
When we call `releaseTo` there are `address(Fund)`, `address(Manager)`, `address(0)`, `Fund.owner`, `msg.sender` and `_new`.

It is possible that all 6 addresses are distinct.

I can elaborate more in the post if you think that would help.

wuestholz replied:
Yeah, maybe it's fine to just include the 0-address in the list. Then the list would contain 6 elements.

ScottWe replied:
I've updated this in response to Jorge's feedback. I think it resolves your concern as well.

wuestholz replied:
Thanks! Sounds good.

wuestholz commented on tutorials/4_arbitrary_clients.md:
"by appealing to the"?

Commit:
Tut(2-4): links, wording, fixed typos, citations.

Commit:
Tut(3): clarified addrs 0, 1, 2

Commit:
Tut(4): clarified addresses in neighbourhood.

Commit:
Tut(2,3,4): using 'cmake' over 'make'.

Commit:
Tut(1): docker setup.

Commit:
Tut(2): links to pregen models.

Commit:
Tut(3): links to pregen models.

Commit:
Tut(4): links to pregen models.

Commit:
Tut(2,3,4): claifications and conclusions.

This responds to Valentin's feedback.

Commit:
Tut(2,3,4): reflects cmake fixes.

Commit:
Tut(Fix;2,3,4): state all tutorials use containers

Removed abs. paths from non-containerized cmd-lines.

Commit:
Tut(5): instrumenting adeq. checks

Commit:
Tut(5,fix): property statement

Commit:
Tut(5,wip): remarks on verification.

Commit:
Tut(5,wip): conclusions.

Commit:
Tut(5,wip): citations.

Commit:
Tut(5): end-to-end pass.

Commit:
Tut(1-4): updated installation

agurfinkel commented on tutorials/5_client_data.md:
broken sentence

agurfinkel commented on tutorials/5_client_data.md:
`_dst` vs `_destination`, be consistent across the document

agurfinkel commented on tutorials/5_client_data.md:
don't take credit for notation. Say `where`, not `we write`

mariachris commented on tutorials/5_client_data.md:
small 'a' in As presented

agurfinkel commented on tutorials/5_client_data.md:
typo

agurfinkel commented on tutorials/5_client_data.md:
Expressed as a regular expression, the monitor is:

mariachris commented on tutorials/5_client_data.md:
constructor -> construct

agurfinkel commented on tutorials/5_client_data.md:
Instead of assuming everyone knows and remembers what *topology* means, describe the graph that you want. Then say that this is called **topology** and refer to local reasoning paper.

agurfinkel replied:
also good to have a picture

Commit:
Tut(1,2,3,4,5): added by lines.

Commit:
Tut(5): local fixes (typos, etc)

Commit:
Tut(5): new terms (dist./rep.)

Commit:
Tut(5): fixed opening tense.

Commit:
Tut(3,4,5): removed sound/complete.

Commit:
Tut(5): clarified parameterization.

The original draft conflated the parameters to local reasoning and the parameters to SmartACE.

Commit:
Tut(5): flipped topology paragraph.

Commit:
Tut(5): described topology diagrams.

Commit:
Tut(5): modal logic and property.

This explains why our property requires msg.sender != _destination. The relation is made to modal logic.

Commit:
Tut(5): conservative usage of 'invariant'.

In the previous tutorial, we used 'invariant' to mean 'compositional invariant', 'inductive invariant', and 'predicate which we wish to prove is (compositionally/inductively) invariant'. To solve this issue, all predicates are now described explicitly.

Commit:
Tut(5): minor fixes.
